analytic_engine/src/instance/open.rs

```diff
@@ -138,71 +139,265 @@ impl Instance {
     }
 
     /// Open the table.
-    pub async fn do_open_table(
+    pub async fn do_open_tables_of_shard(
         self: &Arc<Self>,
-        space: SpaceRef,
-        request: &OpenTableRequest,
-    ) -> Result<Option<TableDataRef>> {
-        if let Some(table_data) = space.find_table_by_id(request.table_id) {
-            return Ok(Some(table_data));
+        context: TablesOfShardContext,
+    ) -> Result<OpenTablesOfShardResult> {
+        let mut shard_opener = ShardOpener::init(
+            context,
+            self.space_store.manifest.clone(),
+            self.space_store.wal_manager.clone(),
+            self.replay_batch_size,
+            self.make_flusher(),
+            self.max_retry_flush_limit,
+        )?;
+
+        shard_opener.open().await
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TablesOfShardContext {
+    /// Shard id
+    pub shard_id: ShardId,
+    /// Table infos
+    pub table_ctxs: Vec<TableContext>,
+}
+
+#[derive(Clone, Debug)]
+pub struct TableContext {
+    pub table_def: TableDef,
+    pub space: SpaceRef,
+}
+
+#[derive(Debug)]
+enum TableOpenState {
+    RecoverTableMeta(RecoverTableMetaContext),
+    RecoverTableData(RecoverTableDataContext),
+    Failed(crate::instance::engine::Error),
+    Success(Option<SpaceAndTable>),
+}
+
+#[derive(Debug)]
+struct RecoverTableMetaContext {
+    table_def: TableDef,
+    space: SpaceRef,
+}
+
+#[derive(Debug)]
+struct RecoverTableDataContext {
+    table_data: TableDataRef,
+    space: SpaceRef,
+}
+
+pub type OpenTablesOfShardResult = HashMap<TableId, Result<Option<SpaceAndTable>>>;
+
+/// Opener for tables of the same shard
+struct ShardOpener {
+    shard_id: ShardId,
+    manifest: ManifestRef,
+    wal_manager: WalManagerRef,
+    states: HashMap<TableId, TableOpenState>,
+    wal_replay_batch_size: usize,
+    flusher: Flusher,
+    max_retry_flush_limit: usize,
+}
+
+impl ShardOpener {
+    fn init(
+        shard_context: TablesOfShardContext,
+        manifest: ManifestRef,
+        wal_manager: WalManagerRef,
+        wal_replay_batch_size: usize,
+        flusher: Flusher,
+        max_retry_flush_limit: usize,
+    ) -> Result<Self> {
+        let mut states = HashMap::with_capacity(shard_context.table_ctxs.len());
+        for table_ctx in shard_context.table_ctxs {
+            let space = &table_ctx.space;
+            let table_id = table_ctx.table_def.id;
+            let state = if let Some(table_data) = space.find_table_by_id(table_id) {
+                TableOpenState::Success(Some(SpaceAndTable::new(space.clone(), table_data)))
+            } else {
+                TableOpenState::RecoverTableMeta(RecoverTableMetaContext {
+                    table_def: table_ctx.table_def,
+                    space: table_ctx.space,
+                })
+            };
```

Comment:
Why does the tables to open have these two states? Which one state is normal? Shall we take the other state as a mistake?